### DISCUSSION TITLE
No executable found for solver `ipopt` on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by wigging** [IDAES/idaes-pse#958] [Q&A] [answered]
I'm trying to install IDAES on the following system where Python is installed via Miniconda.

```
Ubuntu 20.04.5 LTS x86_64
Linux 5.4.0-125-generic
Bash 5.0.17
Intel Xeon E5-2680 v3 (20) @ 2.499GHz
Python 3.10.4
```

Following the steps in the IDAES documentation for Linux, I ran the following commands and didn't notice any errors during the installation.

```
sudo apt-get install libgfortran5 libgomp1 liblapack3 libblas3

conda create --name idaes python

conda activate idaes

conda install --yes -c IDAES-PSE -c conda-forge idaes-pse

idaes get-extensions
```

Example Python script to run.

```python
from pyomo.environ import Objective, Constraint, value, SolverFactory
from pyomo.environ import ConcreteModel, Var

model = ConcreteModel()
model.x = Var()
model.y = Var()

model.obj = Objective(expr=model.x**2 + model.y**2)

model.con = Constraint(expr=model.x + model.y == 1)

status = SolverFactory('ipopt').solve(model, tee=True)  # tee=True shows the solver log

print('x =', value(model.x))
print('y =', value(model.y))
print('obj =', value(model.obj))
```

Errors reported from running the example are shown below.

```
WARNING: Could not locate the 'ipopt' executable, which is required for solver
    ipopt
Traceback (most recent call last):
  File "/home/cades/idaes-examples/example1.py", line 20, in <module>
    status = SolverFactory('ipopt').solve(model, tee=True)  # tee=True shows the solver log
  File "/home/cades/miniconda3/envs/idaes/lib/python3.10/site-packages/pyomo/opt/base/solvers.py", line 513, in solve
    self.available(exception_flag=True)
  File "/home/cades/miniconda3/envs/idaes/lib/python3.10/site-packages/pyomo/opt/solver/shellcmd.py", line 129, in available
    raise ApplicationError(msg % self.name)
pyomo.common.errors.ApplicationError: No executable found for solver 'ipopt'
```

It looks like the `ipopt` solver was not installed. Is there another step I'm supposed to do for Linux?

**Comment by lbianchi-lbl** [comment]
Hi @wigging, can you try using `from idaes.core.solvers import get_solver` instead of `SolverFactory`?

**Reply by wigging** [reply]
I tried the following but get an import error.

```python
from idaes.core.solvers import get_solver

solver = get_solver()
solver.solve(model, tee=True)
```

Here's the error.

```
Traceback (most recent call last):
  File "/home/cades/idaes-examples/example1.py", line 12, in <module>
    from idaes.core.solvers import get_solver
ImportError: cannot import name 'get_solver' from 'idaes.core.solvers' (/home/cades/miniconda3/envs/idaes/lib/python3.10/site-packages/idaes/core/solvers/__init__.py)
```

**Reply by wigging** [reply]
I think this approach requires watertap to be installed. I'm just trying to run IDAES examples right now and have not installed watertap yet.

**Reply by andrewlee94** [reply]
I suspect the problem is the version of IDAES. My guess is that without specifying a version when you did the conda install, you got the last stable version of IDAES (v1.13). We have changed the API since then in preparation for v2.0, and one of those was moving `get_solver`.

**Reply by lbianchi-lbl** [reply]
@wigging, thanks for trying that. I think this might be due to the aforementioned breaking changes on the IDAES side: I used the current import path, which is different from the legacy import path that's available to the version of IDAES with which WaterTAP is (as of now) compatible. I understand that these issues can be confusing in general, and even more so for someone just joining the project during this "pardon our dust" stage. Our current priority is to make the codebase compatible with IDAES 2.0 as soon as possible.

**Comment by andrewlee94** [comment, marked as answer]
@wigging The issue is that the IDAES IPOPT solver is only available through IDAES (for licensing reasons). You can also fix the issue by doing `import idaes` at the beginingg of your script.

**Reply by lbianchi-lbl** [reply]
Also, there are disadvantages to running scripts instantiating solver objects at the module level. You can read more about the recommended ways to do so here: https://watertap.readthedocs.io/en/latest/how_to_guides/how_to_run_models_in_a_py_script.html

**Reply by andrewlee94** [reply]
It is also possible to manually add the IPOPT executable to your system path.

**Reply by wigging** [reply]
@lbianchi-lbl Yes, I'm aware if running things with `__main__` but left that out to shorten the example code.

@andrewlee94 I've never heard of this ipopt solver. Where do I get more information about it?

**Reply by andrewlee94** [reply]
I am referring to the pre-compiled version of IPOPT you get from the `idaes get-extensions` command.

**Reply by wigging** [reply]
@andrewlee94 Doing `import idaes` works. Is this explained somewhere in the docs that you need to `import idaes`?

**Reply by andrewlee94** [reply]
I think it is, but it is buried somewhere and easy to miss.

**Reply by wigging** [reply]
The Using Solvers Without IDAES section in the docs discusses setting the paths with `import idaes`.
https://idaes-pse.readthedocs.io/en/stable/tutorials/getting_started/binaries.html
